modules/concurrency/include/hephaestus/concurrency/spinner.h

```diff
@@ -13,22 +13,37 @@ namespace heph::concurrency {
 
 /// A spinner is a class that spins in a loop calling a user-defined function.
 /// If the function is blocking, the spinner will block the thread.
-/// If the input `rate_hz` is set to a non-zero value, the spinner will call the user-defined function at
-/// the given fixed rate.
+/// If the input `rate_hz` is set to a non-infinite value, the spinner will call the user-defined function at
+/// the given fixed rate. The spinner behavior can be configured using callbacks.
 class Spinner {
 public:
   enum class SpinResult : bool { CONTINUE, STOP };
   using StoppableCallback = std::function<SpinResult()>;
   using Callback = std::function<void()>;
 
-  /// @brief Create a spinner with a stoppable callback. A stoppable callback is a function that returns
-  /// SpinResult::STOP to indicate that the spinner should stop.
-  /// Example: a callback that stops after 10 iterations.
-  explicit Spinner(StoppableCallback&& stoppable_callback, double rate_hz = 0);
+  /// @brief Create a callback for the spinner which never stops.
+  [[nodiscard]] static auto createNeverStoppingCallback(Callback&& callback) -> StoppableCallback;
+
+  struct StateMachineCallbacks {
+    using TransitionCallback = std::function<void()>;
+    using PolicyCallback = std::function<bool()>;
+
+    TransitionCallback init_cb = []() {};       //!< Handles initialization.
+    TransitionCallback spin_once_cb = []() {};  //!< Handles execution.
```

Comment:
should we call this? It is more general meaning, but it is also true that we are in a spinner, so probably spin_once is more correct/clear
```suggestion
    TransitionCallback execute_cb = []() {};  //!< Handles execution.
```